src/ExtranetBundle/objects/TCMSTableEditor/TableEditorExtranetUser.php

```diff
@@ -35,58 +40,111 @@ protected function GetCustomMenuItems()
+        $menuItem->sOnClick = sprintf("window.location.href = '%s'; return false;", $url);
 
-        $this->oMenuItems->AddItem($menuItem);
+        return $menuItem;
     }
 
     /**
      * Login as the currently selected extranet user (permissions needed).
+     * This process works by creating a temporary token to verify the login and
+     * redirecting the user to a route on the domain of the corresponding site
+     * which uses said token to log the user in.
+     *
+     * @see LoginController::loginAction()
      */
-    public function LoginAsExtranetUser()
+    public function LoginAsExtranetUser(): void
     {
-        $cmsUser = TCMSUser::GetActiveUser();
-        if (!$cmsUser || !$cmsUser->oAccessManager || !$cmsUser->oAccessManager->PermitFunction('allow-login-as-extranet-user')) {
+        if (false === $this->isBackendUserLoggedInWithPermission('allow-login-as-extranet-user')) {
             return;
         }
-        $inputFilterUtil = $this->getInputFilterUtil();
-        $userId = $inputFilterUtil->getFilteredGetInput('id');
+
+        $userId = $this->getInputFilterUtil()->getFilteredGetInput('id');
         if (null === $userId) {
             return;
         }
+
+        $extranetUser = $this->loadExtranetUser($userId);
+        $portal = $this->portalForExtranetUser($extranetUser);
+        $this->redirectUserToTokenLoginOnPortal($userId, $portal);
+    }
+
+    private function isBackendUserLoggedInWithPermission(string $permission): bool
+    {
+        $cmsUser = TCMSUser::GetActiveUser();
+
+        return $cmsUser && $cmsUser->oAccessManager && $cmsUser->oAccessManager->PermitFunction($permission);
+    }
+
+    private function loadExtranetUser(string $userId): TdbDataExtranetUser
+    {
         $extranetUserProvider = $this->getExtranetUserProvider();
         $extranetUserProvider->reset();
+
+        /** @var TdbDataExtranetUser $extranetUser */
         $extranetUser = $extranetUserProvider->getActiveUser();
         $extranetUser->Load($userId);
+
+        return $extranetUser;
+    }
+
+    private function portalForExtranetUser(TdbDataExtranetUser $extranetUser): TdbCmsPortal
+    {
         if (empty($extranetUser->fieldCmsPortalId)) {
             $portalList = TdbCmsPortalList::GetList();
             $portalList->GoToStart();
-            $portal = $portalList->Current();
-        } else {
-            $portal = TdbCmsPortal::GetNewInstance($extranetUser->fieldCmsPortalId);
+            return $portalList->Current();
         }
-        $extranetUser->DirectLoginWithoutPassword($extranetUser->GetName(), $portal->id);
-        $url = $this->getPageService()->getLinkToPortalHomePageRelative(array(), $portal);
+
+        return TdbCmsPortal::GetNewInstance($extranetUser->fieldCmsPortalId);
+    }
+
+    private function redirectUserToTokenLoginOnPortal(string $userId, TdbCmsPortal $portal): void
+    {
+        $token = $this->transferTokenService()->createTransferTokenForUser($userId, 120);
```

Comment:
I guess the default time could be shorter. As the user will log in immediately after creating the token, 120 seconds seems a bit long. The token is very sensitive data and if leaked somehow 120 seconds is plenty time for an attacker to exploit.

Reply:
:heavy_check_mark: Lowered it to 20 seconds